### DISCUSSION TITLE
No revisions are not being created

### DISCUSSION THREAD

**Discussion by ibonelli** [outline/outline#3612] [Self Hosting] [answered]
**To Reproduce**

Create a document:

- Wait for more than 5mins and change it
- Check the revisions table in the database
- There will be no new revision for the document

**Expected behavior**

When a document changes (beyond the [debounce time](https://github.com/outline/outline/discussions/2016)) a new revision should be created.

**Screenshots**

![Screenshot_20220531_192944b](https://user-images.githubusercontent.com/7817148/171296324-81fc13cc-fd57-4460-85a2-58f3b51087b0.png)

![Screenshot_20220531_194059](https://user-images.githubusercontent.com/7817148/171296326-0b29e9bc-f4bd-4346-84e6-fc33baf07914.png)

**Outline**

- Install: self hosted
- Version: [v0.64.3](https://github.com/outline/outline/releases/tag/v0.64.3)

**Desktop**

- OS: Heroku
- Browser Chrome
- Version v99

**Comments**

Before upgrading to the latest version I entered [another bug report](https://github.com/outline/outline/issues/3544). After upgrade problem persists. Revisions were created before, but they seem to have stopped at a point. I have no record of which upgrade might have broke it. Maybe there is a new configuration I'm missing? At this time I see nothing on the log that would indicate a deprecation, but I must admit I do not check every time I do an upgrade.

Also looking at the [RevisionsProcessor.ts](https://github.com/outline/outline/blob/main/server/queues/processors/RevisionsProcessor.ts) code I see nothing that would prevent the revisions from being created. The log doesn't report problems.

Help?

**Comment by Limezy** [comment]
Hi, I have tried but couldn't reproduce so far.
Did you activate the collaborative editing ?
Could you elaborate your steps to produce the bug ?
> Create a document:
> Wait for more than 5mins and change it

* Should I publish the document right away ?
* Should I add a title to the document before waiting ?

**Comment by tommoor** [comment]
We definitely need to know if you're using collaborative editing or not – these are totally separate code paths between old/new. Also if you could edit the doc and then post the resulting logs that would be useful, it should show which events are queued and which processors are triggered.

**Comment by ibonelli** [comment]
Hi both, thanks for the quick reply.

Firstly, yes I have collaborative editing on:

![Screenshot_20220601_093334b](https://user-images.githubusercontent.com/7817148/171406559-9fb032de-3927-43cc-b653-c8440dc03572.png)

In terms of log/debugging I modified (same document as before) and I get:

![Screenshot_20220601_093839b](https://user-images.githubusercontent.com/7817148/171407347-cfa0a432-f8be-4a42-a9c4-8d3c7c35742d.png)

So it gets persisted, but that was it. Maybe I need to increase debugging to provide more information? I saw no events or processor information...

**Comment by tommoor** [comment]
Please provide the logs from server startup if you can.

**Comment by ibonelli** [comment]
Sure, no problem. As it is in Heroku, I had to tweak an environment variable to restart it: [heroku.log](https://github.com/outline/outline/files/8814694/heroku.log)
 
Hopefully that's enough? Please let me know if I can provide further information. Thanks!

**Reply by tommoor** [reply]
Are you running the worker process? It's not evident in these logs – it runs by default. Moved to discussions because this is a config error, not a bug.

**Comment by ibonelli** [comment, marked as answer]
Yes, I agree it seems more like a configuration issue.

For what I see in Heroku I have two different instances:

```
web: yarn start --services=web,websockets,collaboration
worker: yarn start --services=worker
```

I see **web** running, but **worker** is limited by the amount of Heroku instances I'm running. So, can I assume running the worker as well should fix the problem?

If that's the case, I guess when I deployed a release with the [Refactoring event processors and service architecture](https://github.com/outline/outline/commit/28aef82af9d2664c13db4ae6aba6fe88476ec2a3) the revisions stopped working due to the lack of the **service worker**. Right?

Thanks for your help!

**Reply by tommoor** [reply]
> the revisions stopped working due to the lack of the service worker. Right?

Yes, revisions are created on the worker now. Lots of functionality runs on the workers, they are a vital part of the architecture 😉

**Reply by almereyda** [reply]
When scaling out Outline, would you have a rule of thumb available for us, in which scenarios you launch more main and worker instances?

**Comment by ibonelli** [comment]
Perfect. thanks a lot for helping me figure this out. Cheers!

**Reply by almereyda** [reply]
Thank you for bringing this up. The above commands can also help to scale other containerised workloads better, e.g. the often mobilised Docker Compose, by separating the two processes into two independently scalable containers.
